File: dotnet/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example3/BaseSupportHandler.cs
namespace ChainOfResponsibilityDesignPattern.Example3;

public abstract class BaseSupportHandler : ISupportHandler
{
    private ISupportHandler? _nextHandler;
    protected string HandlerName { get; set; }

    protected BaseSupportHandler(string handlerName)
    {
        HandlerName = handlerName;
    }

    public ISupportHandler SetNext(ISupportHandler handler)
    {
        _nextHandler = handler;
        return handler;
    }

    public virtual void HandleTicket(SupportTicket ticket)
    {
        if (_nextHandler != null)
        {
            _nextHandler.HandleTicket(ticket);
        }
    }

    protected bool CanHandle(SupportTicket ticket, TicketPriority maxPriority, List<TicketCategory>? categories = null)
    {
        bool priorityMatch = ticket.Priority <= maxPriority;
        bool categoryMatch = categories == null || categories.Contains(ticket.Category);
        
        return priorityMatch && categoryMatch;
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example3/src/BaseSupportHandler.ts
import { ISupportHandler, SupportTicket, TicketCategory, TicketPriority } from "./SupportTicket";

export abstract class BaseSupportHandler implements ISupportHandler {
  private _nextHandler: ISupportHandler | null = null;

  constructor(protected readonly handlerName: string) {}

  setNext(handler: ISupportHandler): ISupportHandler {
    this._nextHandler = handler;
    return handler;
  }

  handleTicket(ticket: SupportTicket): void {
    if (this._nextHandler) {
      this._nextHandler.handleTicket(ticket);
    }
  }

  protected canHandle(
    ticket: SupportTicket,
    maxPriority: TicketPriority,
    categories?: TicketCategory[]
  ): boolean {
    const priorityMatch = ticket.priority <= maxPriority;
    const categoryMatch = !categories || categories.includes(ticket.category);
    return priorityMatch && categoryMatch;
  }
}
